geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/planning/guard/GraduatedQueryGuard.scala

```diff
@@ -0,0 +1,145 @@
+   *
+   * @param ds  data store
+   * @param sft simple feature type
+   */
+  override def init(ds: DataStore, sft: SimpleFeatureType): Unit = {
+    guardLimits = loadConfiguration(sft)
+  }
+
+  /**
+   * Modifies the query in place
+   *
+   * @param query query
+   */
+  override def rewrite(query: Query): Unit = { }
+
+  /**
+   * Hook to allow interception of a query after extracting the query values
+   *
+   * @param strategy query strategy
+   * @return an exception if the query should be stopped
+   */
+  override def guard(strategy: api.QueryStrategy): Option[IllegalArgumentException] = {
+    val intervalsOption: Option[FilterValues[Bounds[ZonedDateTime]]] = strategy.values.collect { case v: TemporalIndexValues => v.intervals }
+    val spatialBoundsOption: Option[Seq[(Double, Double, Double, Double)]] = strategy.values.collect { case v: SpatialIndexValues => v.spatialBounds }
+
+    val spatialExtent = spatialBoundsOption match {
+      case None => 360.0 * 180.0  // Whole world
+      case Some(seq) => seq.map { case  (lx: Double, ly: Double, ux: Double, uy: Double) =>
+        (ux-lx) * (uy-ly)
+      }.sum
+    }
+
+    intervalsOption match {
+      // The None case reflects a query without a temporal extent or a full-table scan on a table with a temporal extent
+      // The full-table scan block should be used to avoid the later case.
+      // Here a none is returned indicating that this guard does not stop the query
+      case None => None
+      case Some(intervals) =>
+        if (!intervals.forall(_.isBoundedBothSides) || intervals.isEmpty) {
+          Some(new IllegalArgumentException("At least one part of the query is temporally unbounded.  Add bounded temporal constraints."))
+        } else {
+          val queryDuration = duration(intervals.values)
+          guardLimits.find( _.sizeLimit >= spatialExtent).flatMap {
+            limit =>
+              if (queryDuration > limit.durationLimit) {
+                Some(new IllegalArgumentException(s"Query of spatial size: $spatialExtent and duration: $queryDuration failed with constraint: $limit"))
+              } else {
+                None
+              }
+          }
+        }
+    }
+  }
+
+  override def close(): Unit = { }
+}
+object GraduatedQueryGuard extends LazyLogging {
+  val ConfigPath = "geomesa.guards.graduated-guard"
+
+  /**
+   *
+   * @param sizeLimit in square degrees
+   * @param durationLimit Maximum duration for a query at or below the spatial size
+   */
+  case class SizeAndDuration(sizeLimit: Int, durationLimit: Duration)
+
+  def loadConfiguration(sft: SimpleFeatureType): Seq[SizeAndDuration]  = {
+    val conf = ConfigFactory.load()
+    try {
+      val guardConfig = conf.getConfigList(s"${ConfigPath}.${sft.getTypeName}")
+      buildLimits(guardConfig)
+    } catch {
+      case e: ConfigException =>
+        logger.error(s"Configuration for GraduatedQueryGuard not available or correct for type ${sft.getTypeName}.  " +
```

Comment:
this is good, but FYI if the init method throws an exception, the framework will catch it and the guard won't even be invoked, so you don't need to ensure it still works if there's an error.